File: backend/models/trainee.go
package models

import "time"

type Trainee struct {
	ID                string             `bson:"_id,omitempty" json:"id"`
	Name              string             `json:"name" bson:"name" binding:"required"`
	PhoneNumber       string             `json:"phone_number" bson:"phone_number" binding:"required"`
	Email             string             `json:"email,omitempty" bson:"email,omitempty"`
	UserID            string             `json:"user_id,omitempty" bson:"user_id,omitempty"`
	DOB               string             `json:"dob" bson:"dob" binding:"required"`
	Gender            string             `json:"gender" bson:"gender" binding:"required"`
	Profession        string             `json:"profession" bson:"profession"`
	Height            float64            `json:"height" bson:"height" binding:"gte=0"`
	TrainerID         string             `json:"trainer_id" bson:"trainer_id" binding:"required"`
	StartDate         string             `json:"start_date" bson:"start_date"`
	MembershipType    string             `json:"membership_type" bson:"membership_type"`
	EmergencyContact  string             `json:"emergency_contact" bson:"emergency_contact"`
	MedicalHistory    string             `json:"medical_history,omitempty" bson:"medical_history,omitempty"`
	SocialHandle      string             `json:"social_handle,omitempty" bson:"social_handle,omitempty"`
	Goals             string             `json:"goals" bson:"goals"`
	Notes             string             `json:"notes,omitempty" bson:"notes,omitempty"`
	ActiveStatus      bool               `json:"active_status" bson:"active_status"`
	Rating            float64            `json:"rating" bson:"rating"`
	RatingCount       int                `json:"rating_count" bson:"rating_count"`
	ProgressMetrics   map[string]float64 `json:"progress_metrics,omitempty" bson:"progress_metrics,omitempty"`
	CreatedAt         time.Time          `json:"created_at" bson:"created_at"`
	UpdatedAt         time.Time          `json:"updated_at" bson:"updated_at"`
	ImageURL          string             `json:"image_url,omitempty" bson:"image_url,omitempty"`
	ActiveSupplements string             `json:"active_supplements" bson:"active_supplements"`

	FitnessTests        []FitnessTest        `json:"fitness_tests,omitempty" bson:"fitness_tests,omitempty"`
	LabTests            []LabTest            `json:"lab_tests,omitempty" bson:"lab_tests,omitempty"`
	HealthQuestionnaire *HealthQuestionnaire `json:"health_questionnaire,omitempty" bson:"health_questionnaire,omitempty"`
}

type FitnessTest struct {
    Date         string `json:"date" bson:"date"`
    Category     string `json:"category" bson:"category"`
    Test         string `json:"test" bson:"test"`
    Result1      string `json:"result1" bson:"result1"`
    Result2      string `json:"result2,omitempty" bson:"result2,omitempty"`
}

type LabTest struct {
    Date     string `json:"date" bson:"date"`         
    Category string `json:"category" bson:"category"` 
    Test     string `json:"test" bson:"test"`         
    Value    string `json:"value" bson:"value"`
}

type HealthQuestionnaire struct {
    Responses map[string]bool `json:"responses" bson:"responses"`
    Comments  string          `json:"comments" bson:"comments"`
}

File: frontend/app/types/trainee.ts
//trainee.ts

export type FitnessTest = {
    date: string;
    category: string;
    test: string;
    result1: string;
    result2: string;
};

export type LabTest = {
    date: string;
    category: string;
    test: string;
    value: string;
};

export type HealthQuestionnaireResponses = {
    heartProblems: boolean;
    bloodPressure: boolean;
    chronicIllness: boolean;
    physicalDifficulty: boolean;
    physicianAdvice: boolean;
    recentSurgery: boolean;
    pregnancy: boolean;
    breathingDifficulty: boolean;
    muscleInjury: boolean;
    diabetes: boolean;
    smoking: boolean;
    obesity: boolean;
    cholesterol: boolean;
    familyHeartProblems: boolean;
    hernia: boolean;
    frequentFalls: boolean;
};

export const questionMapping = {
    heartProblems: "History of heart problems, chest pain, or stroke",
    bloodPressure: "Increased blood pressure",
    chronicIllness: "Any chronic illness or condition",
    physicalDifficulty: "Difficulty in physical exercise",
    physicianAdvice: "Advice from physician not to exercise",
    recentSurgery: "Recent surgery (last 12 months)",
    pregnancy: "Pregnancy (now or within last 3 months)",
    breathingDifficulty: "History of breathing difficulty or lung problems",
    muscleInjury: "Muscle, joint, or back disorder, or any previous injury",
    diabetes: "Diabetes or thyroid condition",
    smoking: "Cigarette smoking habit",
    obesity: "Obesity (more than 20% over ideal body weight)",
    cholesterol: "Increased blood cholesterol",
    familyHeartProblems: "History of heart problems in immediate family",
    hernia: "Hernia or any condition that may be aggravated by lifting weights",
    frequentFalls: "Do you have frequent falls / lose consciousness / balance",
};

export type Trainee = {
    id: string;
    name: string;
    phone_number: string;
    dob: string;
    gender: string;
    profession?: string;
    height: number;
    start_date?: string;
    membership_type?: string;
    emergency_contact?: string;
    medical_history?: string;
    goals?: string;
    notes?: string;
    active_status: boolean;
    social_handle?: string;
    image_url?: string;
    active_supplements?: string;

    fitness_tests?: FitnessTest[];
    lab_tests?: LabTest[];
    health_questionnaire?: {
        responses: HealthQuestionnaireResponses;
        comments?: string;
        date: string;
    };
};






export interface FitnessTestWithUI extends FitnessTest {
    id: number;
    icon: string;
    color: string;
}

export interface LabTestWithUI extends LabTest {
    id: number;
    unit: string;
    normalRange: string;
    icon: string;
    color: string;
}

export interface TestRecord {
    date: string;
    fitnessTests: FitnessTest[];
    labTests: LabTest[];
}

export interface CategoryGroup {
    [key: string]: FitnessTestWithUI[];
}

export interface LabCategoryGroup {
    [key: string]: LabTestWithUI[];
}

export const fitnessTestTemplates: Omit<FitnessTestWithUI, 'date' | 'result1' | 'result2'>[] = [
    { 
        id: 1, 
        category: 'Aerobic Endurance', 
        test: 'Rockport Test', 
        icon: 'directions-run',
        color: '#3B82F6'
    },
    { 
        id: 2, 
        category: 'Muscular Strength', 
        test: '1 Rep Max (Upper)', 
        icon: 'fitness-center',
        color: '#EF4444'
    },
    { 
        id: 3, 
        category: 'Muscular Strength', 
        test: '1 Rep Max (Lower)', 
        icon: 'fitness-center',
        color: '#EF4444'
    },
    { 
        id: 4, 
        category: 'Muscular Endurance', 
        test: 'Ab Crunches/min', 
        icon: 'timer',
        color: '#F97316'
    },
    { 
        id: 5, 
        category: 'Muscular Endurance', 
        test: 'Free Squats/min', 
        icon: 'timer',
        color: '#F97316'
    },
    { 
        id: 6, 
        category: 'Flexibility', 
        test: 'Sit & Reach', 
        icon: 'accessibility',
        color: '#10B981'
    },
    { 
        id: 7, 
        category: 'Flexibility', 
        test: 'Shoulder Mobility', 
        icon: 'accessibility',
        color: '#10B981'
    },
    { 
        id: 8, 
        category: 'Body Composition', 
        test: 'Push-ups', 
        icon: 'trending-up',
        color: '#8B5CF6'
    }
];

    
export const labTestTemplates: Omit<LabTestWithUI, 'date' | 'value'>[] = [
    // Complete Blood Count (CBC) - Essential
    { 
        id: 1, 
        category: 'Complete Blood Count (CBC)', 
        test: 'Hemoglobin', 
        unit: 'g/dL',
        normalRange: '12-16',
        icon: 'opacity',
        color: '#EA580C'
    },
    { 
        id: 2, 
        category: 'Complete Blood Count (CBC)', 
        test: 'White Blood Cells', 
        unit: '/μL',
        normalRange: '4000-11000',
        icon: 'opacity',
        color: '#EA580C'
    },
    
    // Lipid Profile - Heart Health
    { 
        id: 3, 
        category: 'Lipid Profile', 
        test: 'Total Cholesterol', 
        unit: 'mg/dL',
        normalRange: '<200',
        icon: 'favorite',
        color: '#7C3AED'
    },
    { 
        id: 4, 
        category: 'Lipid Profile', 
        test: 'HDL Cholesterol', 
        unit: 'mg/dL',
        normalRange: '>40',
        icon: 'favorite',
        color: '#7C3AED'
    },
    { 
        id: 5, 
        category: 'Lipid Profile', 
        test: 'LDL Cholesterol', 
        unit: 'mg/dL',
        normalRange: '<100',
        icon: 'favorite',
        color: '#7C3AED'
    },
    
    // Kidney Function - Essential
    { 
        id: 6, 
        category: 'Kidney Function Test', 
        test: 'Creatinine', 
        unit: 'mg/dL',
        normalRange: '0.6-1.2',
        icon: 'healing',
        color: '#DC2626'
    },
    
    // Liver Function - Essential
    { 
        id: 7, 
        category: 'Liver Function Test', 
        test: 'ALT (SGPT)', 
        unit: 'U/L',
        normalRange: '7-56',
        icon: 'local-hospital',
        color: '#059669'
    },
    
    // Metabolic - Essential
    { 
        id: 8, 
        category: 'Metabolic Panel', 
        test: 'Fasting Glucose', 
        unit: 'mg/dL',
        normalRange: '70-100',
        icon: 'psychology',
        color: '#0891B2'
    },
    { 
        id: 9, 
        category: 'Metabolic Panel', 
        test: 'HbA1c', 
        unit: '%',
        normalRange: '<5.7',
        icon: 'psychology',
        color: '#0891B2'
    },
    
    // Vitamins - Most Important
    { 
        id: 10, 
        category: 'Vitamins', 
        test: 'Vitamin D', 
        unit: 'ng/mL',
        normalRange: '30-100',
        icon: 'wb-sunny',
        color: '#CA8A04'
    },
    { 
        id: 11, 
        category: 'Vitamins', 
        test: 'Vitamin B12', 
        unit: 'pg/mL',
        normalRange: '200-900',
        icon: 'wb-sunny',
        color: '#CA8A04'
    }
];
